# NumPy で扱う値の型

In [2]:
import numpy as np

In [3]:
ar = np.array([1, 2, 3])
print(ar)

[1 2 3]


In [4]:
ar.dtype

dtype('int64')

integer(整数型) 64bit のメモリを使用している 64/8 = 8バイト

In [5]:
ar2 = np.array([1.0, 2.0, 3.0])
print(ar2)

[1. 2. 3.]


- 小数点以下が 0 の場合は省略して表示される事が標準的
- コンピュータの世界では、 integer(整数) と float(小数)内部的には全く異なるモノである

In [6]:
ar2.dtype

dtype('float64')

float : 浮動小数点数 / 小数点は結果によって右に左に動く事から

64bit :64bit のメモリを使用している 64/8 = 8バイト

## 型変換
- <変換前>.astype('変換後の型')

### 機械学習・深層学習・AI系の処理は非常にメモリを使う
- 計算するメモリの容量が足りるのかどうか？が最優先位なってくる
- 特に深層学習
    - 人間の神経細胞の仕組みを再現したニューラルネットワークを用いた機械学習の１つ
    - 上記を扱う時に最重要となるのは、<u>メモリの節約と高速化</u>が大きな課題になる

- もっと短いbit長でメモリの使用量を節約する
    - 深層学習では、float32 に変換して、学習、推論、実行が行われている

### float変換

- float64 :(倍精度)

- float32 :(単精度)


float64は2倍のメモリを消費します。一部のマシンアーキテクチャでは、float64での操作に時間がかかる場合がある

ただし、float64は、32bitのfloatよりも正確に数値を表すことができる

また、はるかに大きな数を格納することもできる

    - 出力にはどのような精度が必要ですか？
    - 速度はまったく問題ではありませんか？
    - 100万分の1で必要な精度は？

#### 単精度へ変換

In [10]:
ar3 = ar2.astype('float32')
print(ar5)

[1. 2. 3.]


In [11]:
ar3.dtype

dtype('float32')

In [12]:
ar4 = np.array([np.pi])  # pi ： 円周率
print(ar4)

[3.14159265]


In [13]:
ar4.dtype

dtype('float64')

In [14]:
ar5 = ar4.astype('float32')
print(ar5)

[3.1415927]


In [15]:
ar5.dtype

dtype('float32')

#### 単精度へ変換後
- float64
    - ar6 = 3.14159265

- float32
    - ar7 = 3.1415927
    - メモリ量が小さくなった分、四捨五入されて省略される

#### <u>表現できる　bit長を短くして、精度が落ちる</u>
- 正確さの犠牲に、メモリを省略できて、計算速度が上がる
- data sience とか AI の情報処理では、あえて精度を少しだけ犠牲にして大量に高速に学習させる
- あえて単精度に変換する

### 深層学習(Neural Network)
- float16
    - 正確さはいらないから、高速性・小メモリを意識した使用方法が深層学習ではある

#### 半精度への変換
- 実際の数値計算で半精度を使用する事はない
- 機械学習の世界、深層学習の世界では使用する事があるので例として

In [16]:
ar6 = ar5.astype('float16')
print(ar6)

[3.14]


In [17]:
ar6.dtype

dtype('float16')

## まとめ
- object.dtype
    - プロパティで型を参照する
- .astype(変換後)
    - methodの方を変換する

- 整数
    - <u>int32</u> : 通常の整数。１つの値を32bitで表現(4バイト)
    - int64 : 非常に大きな整数を扱う際にこれをあえて使用する事がある(8バイト)

- 浮動小数点数
    - <u>float64(倍精度)</u> : 通常の場合はこの型として扱われる(15桁程度の精度) / 64bit で表現(8バイト消費)
    
    - float32(単精度) : 深層ニューラルネットワークの処理で多く用いられる(精度が低い) / 32bit で表現(４バイト消費)

    - float16(半精度) : 深層ニューラルネットワーク。更に精度が低い(数桁程度の表現) / 16bit で表現(２バイト消費)　※ 使用する事もある。実際は 32bit が使用されている

#### 表現の bit長が短い程、メモリの消費量が小さいの上、計算も高速
- 計算の精度と速度、メモリ使用量などからどのタイプにするかを判断する

### NumPy の定数
- np.pi　　:　円周率
- np.e　　:　ネイピア数(自然対数の底)